### PR TITLE
RUM V3: add context page fields

### LINKED ISSUE #3657: [RUM] Extend context.page for allowing new metrics as part of transaction
Currently RUM agent sends page specific information for all managed transactions as part of the transaction context. eg : 

```js
transaction.context.page =  {
    referer: "blah"
    url: "http://example.com"
}
```

some of the metrics like `Memory`, `Network information(4g, 3g)`, `Hardware concurrency(whether the user is on low/high end device)` can only be added to `context.custom` which feels not so intuitive. I would like to propose we find a common schema for adding these navigation level metrics as part of the transaction. It would help with some ideas like

+  Creating RUM dashboards and filtering for users on low end mobiles with limited memory coming from an `search` page. 

These are some of the metrics I thought we could send as part of the page-context. 

```js
{
  page: {
      referer: document.referrer,
      url: window.location.href,
      networkInfo: {
        effectiveType: '4g',
        rtt: 50,
        downlink: 10,
        saveData: false,
        downlinkMax: Infinity,
        type: 'wifi'
      },
      hardwareConcurrency: 4, // cores
      deviceMemory: 4 (float), // A full list of possible values is as follows: 0.25, 0.5, 1, 2, 4, 8
      servedViaServiceWorker: "yes"  // ["yes", "no", "unsupported/not applicable"] 
    }
} 
```

Does allowing * in the page context solve the use case for RUM since its already used only by the RUM agent ?

### PR DESCRIPTION
**Pull request by jalvz**
## Motivation/summary

Collect more data for RUM analytics.

## Checklist

- [ ] I have signed the [Contributor License Agreement](https://www.elastic.co/contributor-agreement/).
- [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-server/blob/master/CHANGELOG.asciidoc)

I have considered changes for:
- [ ] documentation
- [ ] logging (add log lines, choose appropriate log selector, etc.)
- [ ] metrics and monitoring (create issue for Kibana team to add metrics to visualizations, e.g. [Kibana#44001](https://github.com/elastic/kibana/issues/44001))
- [ ] automated tests (add tests for the code changes, all [**unit** tests](https://github.com/elastic/apm-server/blob/master/TESTING.md) pass locally)
- [ ] telemetry
- [ ] Elasticsearch Service (https://cloud.elastic.co)
- [ ] Elastic Cloud Enterprise (https://www.elastic.co/products/ece)
- [ ] Elastic Cloud on Kubernetes (https://www.elastic.co/elastic-cloud-kubernetes)

## How to test these changes

Send RUM V3 data with new fields, and see them indexed. 
Send RUM V3 data without new fields, and see that they are not rejected.

## Related issues

Closes elastic/apm-server#3657

### CODE REVIEW

File: model/context.go [hunk 1]
```diff
@@ -107,10 +107,37 @@ func truncate(s string) string {
...
+func (ni *NetworkInfo) fields() common.MapStr {
+	fields := common.MapStr{}
+	if ni == nil {
+		return fields
+	}
+	utility.Set(fields, "type", ni.Type)
+	utility.Set(fields, "downlink", ni.Downlink)
+	utility.Set(fields, "downlinkMax", ni.DownlinkMax)
+	utility.Set(fields, "effectiveType", ni.EffectiveType)
+	utility.Set(fields, "roundTripTime", ni.RoundTripTime)
```
Comment (axw) [inline model/context.go:128]: Maybe just "rtt"? It's a pretty well known acronym.

File: model/context.go [hunk 2]
```diff
@@ -107,10 +107,37 @@ func truncate(s string) string {
...
+	utility.Set(fields, "saveData", ni.SaveData)
```
Comment (axw) [inline model/context.go:129]: I don't see "savedata" in the [Network Information API](https://wicg.github.io/netinfo/#networkinformation-interface) spec. Seems to me it's more a property of the request (page) than the network information.

@vigneshshanmugam where does this property come from?

File: model/context.go [hunk 3]
```diff
@@ -107,10 +107,37 @@ func truncate(s string) string {
 	return s
 }
 
+type NetworkInfo struct {
+	EffectiveType *string
+	RoundTripTime *int64
+	Downlink      *int64
+	DownlinkMax   *int64
```
Comment (axw) [inline model/context.go:114]: Downlink could be fractional, so should be a float64:

> The downlink attribute represents the effective bandwidth estimate in megabits per second, rounded to nearest multiple of 25 kilobits per second

Same for downlinkMax.

File: model/modeldecoder/context.go [hunk 4]
```diff
@@ -218,14 +218,39 @@ func decodePage(raw common.MapStr, hasShortFieldNames bool, err error) (*model.P
...
+func decodeNetworkInfo(raw map[string]interface{}, hasShortFieldNames bool) (*model.NetworkInfo, error) {
+	decoder := utility.ManualDecoder{}
+	fieldName := field.Mapper(hasShortFieldNames)
+	networkInfo := &model.NetworkInfo{
+		EffectiveType: decoder.StringPtr(raw, fieldName("effectiveType")),
+		RoundTripTime: decoder.Int64Ptr(raw, fieldName("roundTripTime")),
+		Downlink:      decoder.Int64Ptr(raw, fieldName("downlink")),
+		DownlinkMax:   decoder.Int64Ptr(raw, fieldName("downlinkMax")),
+		SaveData:      decoder.BoolPtr(raw, fieldName("saveData")),
+		Type:          decoder.StringPtr(raw, fieldName("referer")),
```
Comment (axw) [inline model/modeldecoder/context.go:249]:
```suggestion
		Type:          decoder.StringPtr(raw, fieldName("type")),
```

File: model/context.go [hunk 5]
```diff
@@ -107,10 +107,37 @@ func truncate(s string) string {
...
 // Page consists of URL and referer
 type Page struct {
-	URL     *URL
-	Referer *string
+	URL                    *URL
+	Referer                *string
+	NetworkInfo            *NetworkInfo
+	Cores                  *int
+	Memory                 *int
+	ServedViaServiceWorker *string
```
Comment (axw) [inline model/context.go:140]: in https://github.com/elastic/apm-server/issues/3657#issuecomment-615128282 @vigneshshanmugam said we're still exploring whether this should be a boolean or keyword. Has that been clarified since?

File: model/context.go [hunk 6]
```diff
@@ -107,10 +107,37 @@ func truncate(s string) string {
...
+
 // Page consists of URL and referer
 type Page struct {
-	URL     *URL
-	Referer *string
+	URL                    *URL
+	Referer                *string
+	NetworkInfo            *NetworkInfo
+	Cores                  *int
+	Memory                 *int
```
Comment (axw) [inline model/context.go:139]: This should be a float64: https://developer.mozilla.org/en-US/docs/Web/API/Navigator/deviceMemory

Comment (axw) [inline model/context.go:139]: NetworkInfo, Cores, and Memory don't really feel to me like they belong in "Page". They're properties of the device, not the page.

File: model/context.go [hunk 7]
```diff
@@ -203,6 +230,10 @@ func (page *Page) Fields() common.MapStr {
 		utility.Set(fields, "url", page.URL.Original)
 	}
 	utility.Set(fields, "referer", page.Referer)
+	utility.Set(fields, "system.cpu.cores", page.Cores)
+	utility.Set(fields, "system.memory.total", page.Memory)
+	utility.Set(fields, "servedViaServiceWorker", page.ServedViaServiceWorker)
+	utility.Set(fields, "networkInfo", page.NetworkInfo.fields())
```
Comment (axw) [inline model/context.go:236]: As mentioned on https://github.com/elastic/apm-server/issues/3657#issuecomment-615058295, networkInfo feels to me like extensions of the "network" ECS fieldset. Do you see any problems with that?

e.g. networkInfo.type -> network.physical (or something like that; there's precedent in "network.transport")

File: model/context.go [hunk 8]
```diff
@@ -107,10 +107,37 @@ func truncate(s string) string {
...
+}
+
+func (ni *NetworkInfo) fields() common.MapStr {
+	fields := common.MapStr{}
+	if ni == nil {
+		return fields
+	}
+	utility.Set(fields, "type", ni.Type)
+	utility.Set(fields, "downlink", ni.Downlink)
+	utility.Set(fields, "downlinkMax", ni.DownlinkMax)
```
[the remark this answers is not in the document]
Comment (jalvz) [reply model/context.go:126]: ok - but note we haven't been doing that for RUM so far (`timeToFirstByte`, `requestStart`, etc), changing gears now can be more confusing for people dealing with those fields down the line.

File: model/context.go (hunk 1 above)
Comment (jalvz) [reply model/context.go:128]: maybe is not 100% known (particularly for non native speakers)?  what's the harm in `roundTripTime` ?

File: model/context.go (hunk 3 above)
Comment (jalvz) [reply model/context.go:114]: my understand was that rum would not send decimal number anyways. will change..

File: model/context.go (hunk 6 above)
Comment (jalvz) [reply model/context.go:139]: In https://github.com/elastic/apm-server/issues/3657#issuecomment-644658711 I understood that these fields will be queried/used in the context of a page load, so I thought it would be intuitive to set them there, following the original proposal.  
Having a `device` field or similar on the `transaction` doesn't feel any more semantically appropriate to me (a device is not a property of a transaction), and using metricsets as I suggested in the issue would not help querying.

I guess it could be shoehorned in transaction context - just because context means pretty much anything...
Do you have any suggestion in particular?

File: model/context.go (hunk 2 above)
Comment (vigneshshanmugam) [reply model/context.go:129]: You are right, the spec network Information API got updated and it no longer has `saveData` but its supported by chrome and might be added in future so we can spec it and remove usage if not required. 
Screenshot 2020-06-26 at 10 47 47

Comment (vigneshshanmugam) [reply model/context.go:129]: Preview is here - https://s3.amazonaws.com/pr-preview/WICG/netinfo/saveData.html

File: model/context.go (hunk 5 above)
Comment (vigneshshanmugam) [reply model/context.go:140]: It will be string with predefined values. Updated the description with more context.

File: model/context.go (hunk 6 above)
Comment (vigneshshanmugam) [reply model/context.go:139]: Good point Andrew, I am also inclined more towards having it as part of `device` that page. But would it affect/confused with the user_agent.device in any way?

File: docs/spec/rum_v3_context.json [hunk 9]
```diff
@@ -153,6 +153,81 @@
...
+                            "maxLength": 1024
+                        },
+                        "rtt": {
+                            "description": "roundTripTime",
+                            "type": [
+                                "integer",
+                                "null"
+                            ]
+                        },
+                        "dlk": {
```
Comment (vigneshshanmugam) [inline docs/spec/rum_v3_context.json:179]: any reason for three letter combination instead of just `d` ? 
```
downlink -> d
downlinkMax -> dm
saveData -> sd
```

that would align with the camelCasing of the spec.

Comment (jalvz) [reply docs/spec/rum_v3_context.json:179]: d and sd are all already used for duration and started fields, so we would know which one refers too. we could potentially resolve ambiguity by looking at the parent field, but that would entail more complexity than what i am willing to introduce...

File: model/context.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (axw) [reply model/context.go:126]: The motivation behind my request was to move these fields closer to something that we could add to ECS. These network-related fields are more general than timeToFirstByte etc., which are pretty RUM-specific. I'm fine with the latter being non-ECS compliant for the foreseeable future.

[the remark this answers is not in the document]
Comment (jalvz) [reply model/context.go:126]: fare enough, then ill stick basically to those 2 (`downlink_max`, `effective_type`), doesn't look like anything else can be ECS-ified.

File: model/context.go (hunk 7 above)
Comment (jalvz) [reply model/context.go]: no, no problem. but i understand right `network` should still be under `transaction.device`, and not as top level field?

File: docs/spec/rum_v3_context.json (hunk 9 above)
Comment (vigneshshanmugam) [reply docs/spec/rum_v3_context.json:188]: Fair enough, not a big of a deal. I forgot that we dont take depth of parent in to consideration. Let keep it as it is for now.

File: model/context.go (hunk 1 above)
Comment (axw) [reply model/context.go]: There's nothing wrong with "roundTripTime"; I'm arguing that it's more common to write "rtt" than it is "round-trip time" in technical settings. It is of course not known by 100% of people, just as the phrase wouldn't be; but it is a standard technical phrase and acronym in our industry: https://en.wikipedia.org/wiki/Round-trip_delay_time.

Looking through beats, I didn't find any fields with "roundTripTime" or "round_trip_time" in the name. There are several occurrences of fields with "rtt" in the name though:
- https://www.elastic.co/guide/en/beats/filebeat/current/exported-fields-googlecloud.html#_vpcflow_2
- https://www.elastic.co/guide/en/beats/heartbeat/current/exported-fields-tls.html#_rtt_6 (repeats for other protocols)

File: model/context.go (hunk 6 above)
Comment (axw) [reply model/context.go]: > Having a device field or similar on the transaction doesn't feel any more semantically appropriate to me (a device is not a property of a transaction), and using metricsets as I suggested in the issue would not help querying.

Fair point that it's not a property of the transaction. What I was getting at is that this information is not necessarily restricted to page-loads, so let's not overfit to this one requirement. These fields may also be applicable for serverless/lambda.

Here's a more concrete suggestion:

 1. extend our `system` model type with optional fields to report the the number of CPU cores and total memory
 2. add a `network` model type which can be reported in stream metadata
 3. allow `transaction.context.system` which should be merged with the system from stream metadata, like we do with `transaction.context.service`
 4. same for `transaction.context.network`

We could skip 3-4 if we want these properties to be associated with spans, error and metrics as well as transactions.

File: model/context.go (hunk 7 above)
Comment (axw) [reply model/context.go]: No, where I talked about "device" I was just talking concepts. The recorded fields should be top-level.

File: model/context.go [hunk 10]
```diff
@@ -107,12 +108,42 @@ func truncate(s string) string {
 	return s
 }
 
+type NetworkInfo struct {
+	EffectiveType *string
+	RoundTripTime *int64
+	Downlink      *float64
+	DownlinkMax   *float64
+	SaveData      *bool
```
Comment (axw) [inline model/context.go:116]: Can we please separate `SaveData` out? It's not a property of the network like all the other things, it's both network-related and specific to the page-load request.

Maybe this one can be a direct field of `Page`.

File: model/context.go [hunk 11]
```diff
@@ -107,12 +108,42 @@ func truncate(s string) string {
...
 type Page struct {
 	URL     *URL
 	Referer *string
 }
 
+type Device struct {
+	NetworkInfo            *NetworkInfo
+	Cores                  *int
+	Memory                 *float64
+	ServedViaServiceWorker *string
```
Comment (axw) [inline model/context.go:144]: Please move `ServedViaServiceWorker` to `Page`. This is not related to the device, and is page-load specific.

File: docs/spec/rum_v3_network.json [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+            ]
+        },
+        "dlkm": {
+            "description": "downlink_max",
+            "type": [
+                "number",
+                "null"
+            ]
+        },
+        "ph": {
```
Comment (vigneshshanmugam) [inline docs/spec/rum_v3_network.json:39]: To what field, this maps to the Network Info API?

**Comment by axw** [comment]
Thanks for the updates @jalvz, it's looking much better to me now.

I'd like to hold off on this change for the moment, and probably defer to 7.10. I'm concerned that if we push this, we may end up getting the field names wrong (with respect to ECS), and have to rework them and deal with all of the backwards compatibility issues that entails. As far as I'm aware we don't strictly need this for 7.9, so I'd prefer to take a more measured approach.

To take this forward, I propose we:
 1. Move this PR into Draft
 2. Replace https://github.com/elastic/apm-server/issues/3657 with an elastic/apm issue using the [New Metadata/Context Fields issue](https://github.com/elastic/apm/issues/new?labels=agents%2C+poll&template=new-fields-issue.md) template, using the current state of this PR as the basis
 3. Once we're agreed upon that, raise one or more [RFCs](https://github.com/elastic/ecs/tree/master/rfcs) for ECS

**Comment by jalvz** [comment]
> using the current state of this PR as the basis

Just the basis indeed, as the fields have only been spec'ed for RUM V3

File: docs/spec/rum_v3_network.json (hunk 12 above)
Comment (jalvz) [reply docs/spec/rum_v3_network.json:39]: short for "physical layer", equivalent to "type".

**Comment by vigneshshanmugam** [comment]
@axw @jalvz Should we hold on to the implementation from the RUM agent side? Suggestions?

**Comment by axw** [comment]
@vigneshshanmugam I think it would be best not to merge anything yet.

**Comment by axw** [comment]
This is very stale, closing until there's some resolution of adding network details to ECS: https://github.com/elastic/apm-server/issues/3657#issuecomment-788538219
